File: src/main/java/com/sinnerschrader/s2b/accounttool/logic/component/mapping/GroupMapping.java
package com.sinnerschrader.s2b.accounttool.logic.component.mapping;

import com.sinnerschrader.s2b.accounttool.config.ldap.LdapConfiguration;
import com.sinnerschrader.s2b.accounttool.logic.entity.Group;
import com.sinnerschrader.s2b.accounttool.logic.entity.GroupOfNames;
import com.sinnerschrader.s2b.accounttool.logic.entity.PosixGroup;
import com.unboundid.ldap.sdk.SearchResultEntry;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;

/**
 * Converter for Groups from LDAP to Application
 */
public class GroupMapping implements ModelMaping<Group> {

	@Autowired
	private LdapConfiguration ldapConfiguration;

	@Override
	public Group map(SearchResultEntry entry) {
		if (entry == null) return null;
		final List<String> objectClasses = Arrays.asList(entry.getObjectClassValues());
		final String cn = entry.getAttributeValue("cn");
		if (isPosixGroup(objectClasses)) {
			return new PosixGroup(
					entry.getDN(),
					cn,
					entry.getAttributeValueAsInteger("gid"),
					entry.getAttributeValue("description"),
					getGroupClassification(cn),
					entry.getAttributeValues("memberUid"));
		}
		final boolean unique = isGroupOfUniqueNames(objectClasses);
		if (unique || isGroupOfNames(objectClasses)) {
			final String memberAttribute =
					unique
							? Group.GroupType.GroupOfUniqueNames.getMemberAttritube()
							: Group.GroupType.GroupOfNames.getMemberAttritube();

			return new GroupOfNames(
					entry.getDN(),
					cn,
					entry.getAttributeValue("description"),
					unique,
					getGroupClassification(cn),
					entry.getAttributeValues(memberAttribute));
		}
		throw new IllegalArgumentException(
				"Provided result entry is not supported. Please call isCompatible before.");
	}

	private boolean isAdminGroup(String cn) {
		return StringUtils.containsAny(cn, "admins", "administrators")
				|| ldapConfiguration.getAdministrationGroups().contains(cn)
				|| StringUtils.startsWith(cn, ldapConfiguration.getGroupPrefixes().getAdmin());
	}

	private boolean isTechnicalGroup(String cn) {
		return StringUtils.startsWith(cn, ldapConfiguration.getGroupPrefixes().getTechnical());
	}

	private boolean isTeamGroup(String cn) {
		return StringUtils.startsWith(cn, ldapConfiguration.getGroupPrefixes().getTeam());
	}

	private Group.GroupClassification getGroupClassification(String cn) {
		if (isAdminGroup(cn)) {
			return Group.GroupClassification.ADMIN;
		}
		if (isTechnicalGroup(cn)) {
			return Group.GroupClassification.TECHNICAL;
		}
		if (isTeamGroup(cn)) {
			return Group.GroupClassification.TEAM;
		}
		return Group.GroupClassification.UNKNOWN;
	}

	private boolean isGroupOfNames(Collection<String> objectClasses) {
		return objectClasses.contains(Group.GroupType.GroupOfNames.getObjectClass());
	}

	private boolean isGroupOfUniqueNames(Collection<String> objectClasses) {
		return objectClasses.contains(Group.GroupType.GroupOfUniqueNames.getObjectClass());
	}

	private boolean isPosixGroup(Collection<String> objectClasses) {
		return objectClasses.contains(Group.GroupType.Posix.getObjectClass());
	}

	@Override
	public boolean isCompatible(SearchResultEntry entry) {
		if (entry == null) {
			return false;
		}
		final List<String> objectClasses = Arrays.asList(entry.getObjectClassValues());
		return isPosixGroup(objectClasses)
				|| isGroupOfNames(objectClasses)
				|| isGroupOfUniqueNames(objectClasses);
	}
}

File: src/main/java/com/sinnerschrader/s2b/accounttool/logic/component/mapping/GroupMapping.kt
package com.sinnerschrader.s2b.accounttool.logic.component.mapping

import com.sinnerschrader.s2b.accounttool.config.ldap.LdapConfiguration
import com.sinnerschrader.s2b.accounttool.logic.entity.Group
import com.sinnerschrader.s2b.accounttool.logic.entity.Group.GroupClassification.*
import com.sinnerschrader.s2b.accounttool.logic.entity.Group.GroupType
import com.sinnerschrader.s2b.accounttool.logic.entity.GroupOfNames
import com.sinnerschrader.s2b.accounttool.logic.entity.PosixGroup
import com.unboundid.ldap.sdk.SearchResultEntry
import org.apache.commons.lang3.ArrayUtils.EMPTY_STRING_ARRAY
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.Arrays.asList

@Service
class GroupMapping {

    @Autowired
    private lateinit var ldapConfiguration: LdapConfiguration

    fun map(entry: SearchResultEntry?): Group? {
        if (entry == null) return null

        fun SearchResultEntry.str(attributeName: String) = getAttributeValue(attributeName)
        fun SearchResultEntry.int(attributeName: String) = getAttributeValueAsInteger(attributeName)

        with(entry) {
            try {
                val objectClasses = asList(*objectClassValues)
                val cn = str("cn")

                return when {
                    isPosixGroup(objectClasses) ->
                        PosixGroup(
                                dn = dn,
                                cn = cn,
                                gid = int("gid"),
                                description = str("description") ?: "",
                                groupClassification = getGroupClassification(cn),
                                memberIds = asList(*getAttributeValues("memberUid") ?: EMPTY_STRING_ARRAY)
                        )
                    isGroupOfUniqueNames(objectClasses) -> GroupOfNames(
                            dn = dn,
                            cn = cn,
                            description = str("description") ?: "",
                            isUniqueNames = true,
                            groupClassification = getGroupClassification(cn),
                            memberIds = asList(*getAttributeValues(GroupType.GroupOfUniqueNames.memberAttritube) ?: EMPTY_STRING_ARRAY)
                    )
                    isGroupOfNames(objectClasses) -> GroupOfNames(
                            dn = dn,
                            cn = cn,
                            description = str("description") ?: "",
                            isUniqueNames = false,
                            groupClassification = getGroupClassification(cn),
                            memberIds = asList(*getAttributeValues(GroupType.GroupOfNames.memberAttritube) ?: EMPTY_STRING_ARRAY)
                    )
                    else ->
                        throw IllegalArgumentException("Provided result entry is not supported. Please call isCompatible before.")
                }
            } catch (e: Exception) {
                LOG.error("failed to map: " + dn, e)
                return null
            }
        }
    }

    private fun isAdminGroup(cn: String) =
            StringUtils.containsAny(cn, "admins", "administrators") ||
                    ldapConfiguration.permissions.ldapAdminGroup == cn ||
                    cn.startsWith(ldapConfiguration.groupPrefixes.admin)

    private fun isTechnicalGroup(cn: String) = cn.startsWith(ldapConfiguration.groupPrefixes.technical)
    private fun isTeamGroup(cn: String) = cn.startsWith(ldapConfiguration.groupPrefixes.team)

    private fun getGroupClassification(cn: String) =
            when {
                isAdminGroup(cn) -> ADMIN
                isTechnicalGroup(cn) -> TECHNICAL
                isTeamGroup(cn) -> TEAM
                else -> UNKNOWN
            }

    private fun isGroupOfNames(objectClasses: Collection<String>) = objectClasses.contains(GroupType.GroupOfNames.objectClass)
    private fun isGroupOfUniqueNames(objectClasses: Collection<String>) = objectClasses.contains(GroupType.GroupOfUniqueNames.objectClass)
    private fun isPosixGroup(objectClasses: Collection<String>) = objectClasses.contains(GroupType.Posix.objectClass)

    fun isCompatible(entry: SearchResultEntry?): Boolean {
        if (entry == null) return false

        val objectClasses = asList(*entry.objectClassValues)
        return isPosixGroup(objectClasses) ||
                isGroupOfNames(objectClasses) ||
                isGroupOfUniqueNames(objectClasses)
    }

    fun map(entries: List<SearchResultEntry>) = entries.filter { isCompatible(it) }.mapNotNull { map(it) }.sorted()

    companion object {
        private val LOG = LoggerFactory.getLogger(GroupMapping::class.java)
    }
}
